drivers/sensor/lsm6dsr/lsm6dsr_trigger.c

```diff
@@ -0,0 +1,316 @@
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+
+	__ASSERT_NO_MSG(trig->type == SENSOR_TRIG_DATA_READY);
+
+	if (!cfg->trig_enabled) {
+		LOG_ERR("trigger_set op not supported");
+		return -ENOTSUP;
+	}
+
+	if (trig->chan == SENSOR_CHAN_ACCEL_XYZ) {
+		lsm6dsr->handler_drdy_acc = handler;
+		lsm6dsr->trig_drdy_acc = trig;
+		if (handler) {
+			return lsm6dsr_enable_xl_int(dev, LSM6DSR_EN_BIT);
+		} else {
+			return lsm6dsr_enable_xl_int(dev, LSM6DSR_DIS_BIT);
+		}
+	} else if (trig->chan == SENSOR_CHAN_GYRO_XYZ) {
+		lsm6dsr->handler_drdy_gyr = handler;
+		lsm6dsr->trig_drdy_gyr = trig;
+		if (handler) {
+			return lsm6dsr_enable_g_int(dev, LSM6DSR_EN_BIT);
+		} else {
+			return lsm6dsr_enable_g_int(dev, LSM6DSR_DIS_BIT);
+		}
+	}
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+	else if (trig->chan == SENSOR_CHAN_DIE_TEMP) {
+		lsm6dsr->handler_drdy_temp = handler;
+		lsm6dsr->trig_drdy_temp = trig;
+		if (handler) {
+			return lsm6dsr_enable_t_int(dev, LSM6DSR_EN_BIT);
+		} else {
+			return lsm6dsr_enable_t_int(dev, LSM6DSR_DIS_BIT);
+		}
+	}
+#endif
+
+	return -ENOTSUP;
+}
+
+/**
+ * lsm6dsr_handle_interrupt - handle the drdy event
+ * read data and call handler if registered any
+ */
+static void lsm6dsr_handle_interrupt(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm6dsr_status_reg_t status;
+
+	while (1) {
+		if (lsm6dsr_status_reg_get(ctx, &status) < 0) {
+			LOG_DBG("failed reading status reg");
+			return;
+		}
+
+		if ((status.xlda == 0) && (status.gda == 0)
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+			&& (status.tda == 0)
+#endif
+		) {
+			break;
+		}
+
+		if ((status.xlda) && (lsm6dsr->handler_drdy_acc != NULL)) {
+			lsm6dsr->handler_drdy_acc(dev, lsm6dsr->trig_drdy_acc);
+		}
+
+		if ((status.gda) && (lsm6dsr->handler_drdy_gyr != NULL)) {
+			lsm6dsr->handler_drdy_gyr(dev, lsm6dsr->trig_drdy_gyr);
+		}
+
+		if (IS_ENABLED(CONFIG_LSM6DSR_ENABLE_TEMP)) {
+			if ((status.tda) && (lsm6dsr->handler_drdy_temp != NULL)) {
+				lsm6dsr->handler_drdy_temp(dev, lsm6dsr->trig_drdy_temp);
+			}
+		}
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy,
+					GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static inline void handle_irq(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy, GPIO_INT_DISABLE);
+
+	if (IS_ENABLED(CONFIG_LSM6DSR_TRIGGER_OWN_THREAD)) {
+		k_sem_give(&lsm6dsr->gpio_sem);
+	} else if (IS_ENABLED(CONFIG_LSM6DSR_TRIGGER_GLOBAL_THREAD)) {
+		k_work_submit(&lsm6dsr->work);
+	}
+}
+
+static void lsm6dsr_gpio_callback(const struct device *dev,
+				    struct gpio_callback *cb, uint32_t pins)
+{
+	struct lsm6dsr_data *lsm6dsr =
+		CONTAINER_OF(cb, struct lsm6dsr_data, gpio_cb);
+
+	ARG_UNUSED(pins);
+
+	handle_irq(lsm6dsr->dev);
+}
+
+#ifdef CONFIG_LSM6DSR_TRIGGER_OWN_THREAD
```

Comment:
`#ifdef ... #endif` can be removed

Reply:
The items in the data struct that are only needed for trigger support or when using a thread vs workqueue won't be present.  So even though `lsm6dsr_thread()` will be unused when not using a thread it will still fail to compile because `gpio_sem` is not part of the data struct.

Reply:
I noticed this to be the case, for `CONFIG_LSM6DSR_ENABLE_TEMP` since `int16_t temp_sample;` is not defined if using `IS_ENABLED`.